program_xdp.go

```diff
@@ -61,27 +78,49 @@ func newXdpProgram(name, license string, bytecode []byte) Program {
 	}
 }
 
+// Attach attaches eBPF(XDP) program to network interface.
+// There are 2 possible ways to do that:
+//
+// 1. Pass interface name as parameter, e.g.
+//    xdpProgram.Attach("eth0")
+//
+// 2. Using XdpAttachParams structure:
+//    xdpProgram.Attach(
+//			&XdpAttachParams{Mode: XdpAttachModeSkb, Interface: "eth0"
+//    })
 func (p *xdpProgram) Attach(data interface{}) error {
-	ifname, ok := data.(string)
-	if !ok {
-		return fmt.Errorf("Interface name as string expected, got %T", data)
+	var ifaceName string
+	var attachMode = XdpAttachModeDrv
+
+	switch x := data.(type) {
+	case string:
+		ifaceName = x
+	case *XdpAttachParams:
+		ifaceName = x.Interface
+		attachMode = x.Mode
+	default:
+		return fmt.Errorf("%T is not supported for Attach()", data)
 	}
+
 	// Lookup interface by given name, we need to extract iface index
-	iface, err := netlink.LinkByName(ifname)
+	link, err := netlink.LinkByName(ifaceName)
 	if err != nil {
 		// Most likely no such interface
 		return fmt.Errorf("LinkByName() failed: %v", err)
 	}
 
-	err = netlink.LinkSetXdpFd(iface, p.fd)
-	if err != nil {
+	// Attach program
+	if err := netlink.LinkSetXdpFdWithFlags(link, p.fd, int(attachMode)); err != nil {
 		return fmt.Errorf("LinkSetXdpFd() failed: %v", err)
 	}
-	p.ifname = ifname
+
+	p.ifname = ifaceName
 
 	return nil
 }
 
+// Detach detaches program from network interface
+// Must be previously attached by Attach() call.
 func (p *xdpProgram) Detach() error {
```

Comment:
if XDP was attached in SKB mode - it must be detached with the same mode. otherwise it would be an error

Reply:
It seems work well - this case is covered by integration test:
```go
	err = xdp0.Detach()
	ts.NoError(err)
```

Currently it uses `lo` for tests, but, it also works on real interface (just tried)

Reply:
Got it, fixed.